[cocoa] Block response from localhost for mixed content if not loopback

sysrqb opened the pull request:
#### 40274a8fb39df8ee3974a16846a5bb00780253c8
<pre>
[cocoa] Block response from localhost for mixed content if not loopback
<a href="https://bugs.webkit.org/show_bug.cgi?id=272955">https://bugs.webkit.org/show_bug.cgi?id=272955</a>
<a href="https://rdar.apple.com/126729658">rdar://126729658</a>

Reviewed by Brent Fulgham.

In general, localhost could resolve to any IP address, but the expectation is
that it resolves to either 127.0.0.1 as an IPv4 and [::1] as IPv6. This change
looks at the resolved IP address and blocks the response if we are loading a
subresource from localhost over plaintext HTTP and the main document was loaded
with HTTPS (and it wasn&apos;t from localhost).

* Source/WebCore/platform/network/DNS.cpp:
(WebCore::IPAddress::isLoopback const):
* Source/WebCore/platform/network/DNS.h:
* Source/WebKit/NetworkProcess/NetworkDataTask.cpp:
(WebKit::NetworkDataTask::didReceiveResponse):
* Source/WebKit/NetworkProcess/NetworkDataTask.h:
* Source/WebKit/NetworkProcess/NetworkDataTaskBlob.cpp:
(WebKit::NetworkDataTaskBlob::dispatchDidReceiveResponse):
* Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp:
(WebKit::NetworkDataTaskDataURL::didDecodeDataURL):
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::didReceiveResponse):
* Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.cpp:
(WebKit::NetworkDataTaskCurl::invokeDidReceiveResponse):
* Source/WebKit/NetworkProcess/soup/NetworkDataTaskSoup.cpp:
(WebKit::NetworkDataTaskSoup::dispatchDidReceiveResponse):

Canonical link: <a href="https://commits.webkit.org/278214@main">https://commits.webkit.org/278214@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/174680ace0508f7027bfe05b34c808f8dc0f6333

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49762 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29050 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/1951 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53005 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/440 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/35130 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40612 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51861 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26649 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42832 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21723 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/24081 "Passed tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8133 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/43087 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/46056 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54587 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/49259 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24854 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47996 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26115 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42957 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47029 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26968 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/56743 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7185 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25842 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11656 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

sysrqb commented on Source/WebCore/platform/network/DNS.cpp:
```diff
@@ -80,6 +80,19 @@ bool IPAddress::containsOnlyZeros() const
+        [] (const struct in_addr& address) {
+        return address.s_addr == INADDR_LOOPBACK;
+    }, [] (const struct in6_addr& address) {
+        return !memcmp(&address.s6_addr, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\1", sizeof(address.s6_addr));
```
Any suggestion for making this more reasonable?

brentfulgham replied:
Could you use `in6addr_loopback` ?

sysrqb commented on Source/WebCore/platform/network/DNS.cpp:
```diff
@@ -80,6 +80,19 @@ bool IPAddress::containsOnlyZeros() const
     }), m_address);
 }
 
+bool IPAddress::isLoopback() const
```
This incorporates review feedback from #27270

brentfulgham reviewed (approved):
Looks good! r=me

brentfulgham commented on Source/WebKit/NetworkProcess/NetworkDataTask.cpp:
```diff
@@ -160,6 +160,20 @@ void NetworkDataTask::didReceiveResponse(ResourceResponse&& response, Negotiated
+            completionHandler(PolicyAction::Ignore);
+            cancel();
+            if (m_client)
+                m_client->didCompleteWithError({ String(), 0, url, "Cancelled load from '" + url.stringCenterEllipsizedToLength() + "' because localhost did not resolve to a loopback address." });
```
Do we have a test case for this? Either WPT or our own? If not, we should create one.

sysrqb replied:
Not easily because we need to control what `localhost` resolves to. Maybe we can add some plumbing to override the resolved address with something else. I'll work on that in a follow up.

sysrqb committed:
[cocoa] Block response from localhost for mixed content if not loopback
https://bugs.webkit.org/show_bug.cgi?id=272955
rdar://126729658

Reviewed by Brent Fulgham.

In general, localhost could resolve to any IP address, but the expectation is
that it resolves to either 127.0.0.1 as an IPv4 and [::1] as IPv6. This change
looks at the resolved IP address and blocks the response if we are loading a
subresource from localhost over plaintext HTTP and the main document was loaded
with HTTPS (and it wasn't from localhost).

* Source/WebCore/platform/network/DNS.cpp:
(WebCore::IPAddress::isLoopback const):
* Source/WebCore/platform/network/DNS.h:
* Source/WebKit/NetworkProcess/NetworkDataTask.cpp:
(WebKit::NetworkDataTask::didReceiveResponse):
* Source/WebKit/NetworkProcess/NetworkDataTask.h:
* Source/WebKit/NetworkProcess/NetworkDataTaskBlob.cpp:
(WebKit::NetworkDataTaskBlob::dispatchDidReceiveResponse):
* Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp:
(WebKit::NetworkDataTaskDataURL::didDecodeDataURL):
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::didReceiveResponse):
* Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.cpp:
(WebKit::NetworkDataTaskCurl::invokeDidReceiveResponse):
* Source/WebKit/NetworkProcess/soup/NetworkDataTaskSoup.cpp:
(WebKit::NetworkDataTaskSoup::dispatchDidReceiveResponse):

Canonical link: https://commits.webkit.org/278214@main
